### PR TITLE
0.2.0 (WIP)

### PR DESCRIPTION
**Pull request by 0x00b1**
Features

- [ ] One-dimensional models
- [ ] Three-dimensional models
- [ ] Time-distributed one-dimensional blocks and models
- [ ] Time-distributed two-dimensional blocks and models
- [ ] Time-distributed three-dimensional blocks and models
- [ ] Group normalization layer
- [ ] Upsample layer
- [ ] Two-dimensional feature pyramids
- [ ] Updated benchmark suite
- [ ] Weights for two-dimensional ResNet-50

Depreciations

- [ ] There’s a few. I’ll update this soon.

To-do

- [ ] Upsample layer tests
- [ ] `resize` tests
- [ ] CNTK `resize` implementation
- [ ] Theano `resize` implementation
- [ ] Update CIFAR-10 benchmark hyperparameters
- [ ] Update CIFAR-100 benchmark hyperparameters

### CODE REVIEW

**Review by hgaiser — COMMENTED** [review]
Hmm why remove the `use_bias` arguments? That breaks compatibility with the pretrained Caffe weights. I get why you'd remove them from conv1, apparently the bias is used in that layer, but not in the rest?

https://github.com/KaimingHe/deep-residual-networks/blob/master/prototxt/ResNet-50-deploy.prototxt#L69

File: keras_resnet/blocks/_1d.py [hunk 1]
```diff
@@ -29,18 +25,48 @@ def basic_1d(filters, stage=0, block=0, kernel_size=3, numerical_name=False, str
 
     :param kernel_size: size of the kernel
 
-    :param numerical_name: if true, uses numbers to represent blocks instead of chars (ResNet{101, 152, 200})
```
Comment (hgaiser) [inline keras_resnet/blocks/_1d.py]: Why remove the numerical_names? Did you find another way to fix this?

File: keras_resnet/blocks/_1d.py [hunk 2]
```diff
@@ -61,17 +87,17 @@ def basic_1d(filters, stage=0, block=0, kernel_size=3, numerical_name=False, str
 
     def f(x):
         y = keras.layers.ZeroPadding1D(padding=1, name="padding{}{}_branch2a".format(stage_char, block_char))(x)
-        y = keras.layers.Conv1D(filters, kernel_size, strides=stride, use_bias=False, name="res{}{}_branch2a".format(stage_char, block_char), **parameters)(y)
-        y = keras_resnet.layers.BatchNormalization(axis=axis, epsilon=1e-5, freeze=freeze_bn, name="bn{}{}_branch2a".format(stage_char, block_char))(y)
+        y = keras.layers.Conv1D(filters, kernel_size, strides=stride, name="res{}{}_branch2a".format(stage_char, block_char), **convolution_kwargs)(y)
+        y = keras.layers.BatchNormalization(axis=axis, epsilon=1e-5, name="bn{}{}_branch2a".format(stage_char, block_char), **batch_normalization_kwargs)(y)
```
Comment (hgaiser) [inline keras_resnet/blocks/_1d.py]: Shouldn't this have some `training=False` argument?

Comment (0x00b1) [reply keras_resnet/blocks/_1d.py]: Yep. `**batch_normalization_kwargs`. The logic is here:

https://github.com/broadinstitute/keras-resnet/pull/40/files#diff-45a7689737303b65e69187734a1fd48fR59

File: keras_resnet/blocks/_1d.py (hunk 1 above)
Comment (0x00b1) [reply keras_resnet/blocks/_1d.py]: It’s still a WIP, but I want to standardize names (using the names from KaimingHe/deep-residual-networks) since I’m going to start uploading a suite of weights.

**Comment by 0x00b1** [comment]
@hgaiser Re: `use_bias`. That’s a mistake. I’ll fix shortly. 😆

**Comment by 0x00b1** [comment]
@hgaiser And thanks for the feedback!

File: keras_resnet/blocks/_1d.py (hunk 1 above)
Comment (hgaiser) [reply keras_resnet/blocks/_1d.py]: iirc the issue was that the names of KaimingHe weren't really following a straight set of rules. In one network he uses `a/b/c/...` and in the next he uses `a/b1/b2/b3/b4/...`. How do you plan to handle this?

**Review by hgaiser — COMMENTED** [review]
I think the `BatchNormalization` layer requires a [`training=False`](https://github.com/keras-team/keras/blob/master/keras/layers/normalization.py#L130) in the call, otherwise it will update its weights.

Also, all ZeroPadding layers seem to be 2D. Copy paste error probably?

File: keras_resnet/__init__.py [hunk 3]
```diff
@@ -1,5 +1 @@
 from . import layers
-
-custom_objects = {
-    'BatchNormalization': layers.BatchNormalization,
```
[the remark this answers is not in the document]
Comment (0x00b1) [reply keras_resnet/__init__.py]: Yep. It’s 2.1.3. I’ve tested it and it works. There’s also a test [here](https://github.com/keras-team/keras/blob/master/tests/keras/layers/normalization_test.py#L187-L223).

I also started merging the GroupNormalization layer. I’ve seen _ dramatic_ image segmentation and object detection improvements when `batch_size` is 1 or 2. 😙

**Comment by 0x00b1** [comment]
> @hgaiser I think the BatchNormalization layer requires a training=False in the call, otherwise it will update its weights.

I missed this. I assumed it was in the constructor, but you’re right it’s in the call. Thanks, dude!
